File: apps/ts/src/graphs/registry.ts
/**
 * Graph registry — dispatches graph_id to the correct graph factory.
 *
 * This module provides a Map-based registry for resolving graph factory
 * functions from an assistant's `graph_id`. The server layer uses this
 * instead of hard-coding imports to a specific graph package.
 *
 * Built-in graphs register themselves at module load time via lazy imports.
 * The Map-based design is intentionally extensible — a future plugin loader
 * can call `registerGraph()` to add new graph types at startup without
 * modifying this module.
 *
 * Supported built-in graph IDs:
 *
 * - `"agent"` — react-agent factory (ReAct agent, default)
 *
 * Unknown `graph_id` values fall back to the default (`"agent"`).
 *
 * Usage:
 *
 *   import { resolveGraphFactory } from "./registry";
 *
 *   const buildGraph = resolveGraphFactory(assistant.graph_id);
 *   const compiled = await buildGraph(config, { checkpointer });
 *
 * Extending with a custom graph:
 *
 *   import { registerGraph } from "./registry";
 *
 *   registerGraph("my_custom", myCustomGraphFactory);
 *
 * Reference: apps/python/src/graphs/registry.py
 */

import { DEFAULT_GRAPH_ID, type GraphFactory } from "./types";

// ---------------------------------------------------------------------------
// Registry storage
// ---------------------------------------------------------------------------

/**
 * Maps graph_id → factory function.
 *
 * Lazy-loaded factories are wrapped in a closure that imports the module
 * on first invocation, matching the Python `_lazy_import` pattern.
 */
const _GRAPH_REGISTRY = new Map<string, GraphFactory>();

// ---------------------------------------------------------------------------
// Lazy import helper
// ---------------------------------------------------------------------------

/**
 * Create a lazy-loading wrapper around a dynamic import.
 *
 * The actual module is not loaded until the factory is first called.
 * This avoids loading all graph packages at startup — only the graph
 * actually requested by the assistant triggers its import.
 *
 * Mirrors Python's `_lazy_import(module_path, attribute)`.
 *
 * @param modulePath - Path to the module (relative or package path).
 * @param attribute - Name of the exported factory function (default: "graph").
 * @returns A GraphFactory that lazily imports and delegates to the real factory.
 */
function lazyImport(modulePath: string, attribute: string = "graph"): GraphFactory {
  let cached: GraphFactory | null = null;

  const wrapper: GraphFactory = async (
    config: Record<string, unknown>,
    options?,
  ) => {
    if (cached === null) {
      // Dynamic import — resolved relative to this module at runtime.
      const module = await import(modulePath);
      const factory = module[attribute];
      if (typeof factory !== "function") {
        throw new Error(
          `lazyImport: ${modulePath}.${attribute} is not a function (got ${typeof factory})`,
        );
      }
      cached = factory as GraphFactory;
    }
    return cached(config, options);
  };

  // Preserve a human-readable name for debugging (mirrors Python's __qualname__).
  Object.defineProperty(wrapper, "name", {
    value: `lazy(${modulePath}.${attribute})`,
  });

  return wrapper;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Register a graph factory under a given `graph_id`.
 *
 * Calling this with an ID that is already registered will overwrite
 * the existing factory. Use `resetRegistry()` in tests to restore
 * the built-in registrations.
 *
 * @param graphId - Unique identifier (e.g. `"agent"`, `"research_agent"`).
 * @param factory - An async callable matching the `GraphFactory` signature.
 * @throws Error if `graphId` is empty or `factory` is not a function.
 *
 * @example
 *   registerGraph("my_graph", myGraphFactory);
 */
export function registerGraph(graphId: string, factory: GraphFactory): void {
  if (!graphId) {
    throw new Error("registerGraph: graphId must be a non-empty string");
  }
  if (typeof factory !== "function") {
    throw new Error(
      `registerGraph(${JSON.stringify(graphId)}): factory must be a function`,
    );
  }
  _GRAPH_REGISTRY.set(graphId, factory);
}

/**
 * Register a graph factory with lazy loading (imported on first use).
 *
 * This is the recommended approach for built-in graphs to keep startup fast.
 * The module is only imported when the factory is first called.
 *
 * @param graphId - Unique identifier (e.g. `"agent"`).
 * @param modulePath - Path for dynamic `import()` (e.g. `"./react-agent"`).
 * @param attribute - Name of the factory export (default: `"graph"`).
 * @throws Error if `graphId` is empty.
 *
 * @example
 *   registerGraphLazy("agent", "./react-agent", "graph");
 */
export function registerGraphLazy(
  graphId: string,
  modulePath: string,
  attribute: string = "graph",
): void {
  if (!graphId) {
    throw new Error("registerGraphLazy: graphId must be a non-empty string");
  }
  if (!modulePath) {
    throw new Error(
      `registerGraphLazy(${JSON.stringify(graphId)}): modulePath must be a non-empty string`,
    );
  }
  _GRAPH_REGISTRY.set(graphId, lazyImport(modulePath, attribute));
}

/**
 * Resolve a graph factory function from a `graph_id` string.
 *
 * Resolution order:
 * 1. Exact match in registry → return it.
 * 2. Unknown ID → log warning, fall back to `"agent"` (DEFAULT_GRAPH_ID).
 * 3. Null/undefined → fall back to `"agent"`.
 *
 * @param graphId - The assistant's `graph_id` field. Null/undefined and
 *   unrecognised values fall back to `"agent"`.
 * @returns A GraphFactory ready to be called with config.
 * @throws Error if the registry has no `"agent"` entry (should never happen
 *   in normal operation since built-ins are registered at module load).
 *
 * @example
 *   const factory = resolveGraphFactory("agent");
 *   const agent = await factory(config, { checkpointer });
 */
export function resolveGraphFactory(graphId?: string | null): GraphFactory {
  const effectiveId = graphId || DEFAULT_GRAPH_ID;

  const factory = _GRAPH_REGISTRY.get(effectiveId);
  if (factory !== undefined) {
    return factory;
  }

  // Unknown graph_id — fall back to default with warning.
  if (effectiveId !== DEFAULT_GRAPH_ID) {
    console.warn(
      `[graph-registry] Unknown graph_id="${effectiveId}" — falling back to "${DEFAULT_GRAPH_ID}"`,
    );
    const defaultFactory = _GRAPH_REGISTRY.get(DEFAULT_GRAPH_ID);
    if (defaultFactory !== undefined) {
      return defaultFactory;
    }
  }

  // Last resort: registry has no default entry.
  throw new Error(
    `[graph-registry] No factory registered for "${DEFAULT_GRAPH_ID}". ` +
      "Ensure the built-in graphs are registered at startup.",
  );
}

/**
 * Return the list of all registered graph IDs.
 *
 * Useful for the `/info` endpoint, assistant validation, and debugging.
 *
 * @returns Sorted list of graph ID strings.
 */
export function getAvailableGraphIds(): string[] {
  return Array.from(_GRAPH_REGISTRY.keys()).sort();
}

/**
 * Check whether a graph ID is registered.
 *
 * @param graphId - The graph ID to check.
 * @returns `true` if registered, `false` otherwise.
 */
export function isGraphRegistered(graphId: string): boolean {
  return _GRAPH_REGISTRY.has(graphId);
}

/**
 * Clear the registry and re-register built-in graphs.
 *
 * **For testing only.** Restores the registry to its initial state
 * (only built-in graphs registered via lazy import).
 */
export function resetRegistry(): void {
  _GRAPH_REGISTRY.clear();
  _registerBuiltins();
}

// ---------------------------------------------------------------------------
// Built-in graph registration (lazy imports — no packages loaded at startup)
// ---------------------------------------------------------------------------

/**
 * Register the built-in graphs. Called at module load and by `resetRegistry()`.
 *
 * Each built-in is registered lazily so the actual graph module is only
 * imported when first requested. This keeps server startup fast.
 */
function _registerBuiltins(): void {
  registerGraphLazy("agent", "./react-agent/index", "graph");
}

// Auto-register built-ins at module load time.
_registerBuiltins();

File: apps/python/src/graphs/registry.py
"""Graph registry — dispatches to the correct graph factory.

This module provides a dict-based registry for resolving graph factory
functions from an assistant's ``graph_id``.  The server layer uses this
instead of hard-coding imports to a specific graph package.

Built-in graphs register themselves at import time.  The dict-based
design is intentionally extensible — a future BPMN-to-LangGraph compiler
or plugin loader can call :func:`register_graph` to add new graph types
at startup without modifying this module.

Supported built-in graph IDs:

- ``"agent"`` — :func:`graphs.react_agent.graph` (ReAct agent, default)
- ``"research_agent"`` — :func:`graphs.research_agent.graph`
  (two-phase parallel research with human-in-the-loop)

Unknown ``graph_id`` values fall back to the default (``"agent"``).

Usage::

    from graphs.registry import resolve_graph_factory

    build_graph = resolve_graph_factory(assistant.graph_id)
    compiled = await build_graph(config, checkpointer=cp, store=st)

Extending with a custom graph::

    from graphs.registry import register_graph

    async def my_custom_graph(config, *, checkpointer=None, store=None):
        ...
        return compiled_graph

    register_graph("my_custom", my_custom_graph)
"""

from __future__ import annotations

import logging
from typing import Any, Callable, Coroutine

logger = logging.getLogger(__name__)

# Type alias for an async graph factory function.
# Signature: (config, *, checkpointer=None, store=None) -> compiled graph
GraphFactory = Callable[..., Coroutine[Any, Any, Any]]

# Default graph_id used when the assistant doesn't specify one or the
# value is not recognised.
DEFAULT_GRAPH_ID = "agent"

# ---------------------------------------------------------------------------
# Registry storage
# ---------------------------------------------------------------------------

_GRAPH_REGISTRY: dict[str, GraphFactory | str] = {}
"""Maps graph_id → factory function or dotted import path (lazy)."""


# ---------------------------------------------------------------------------
# Lazy import helper
# ---------------------------------------------------------------------------


def _lazy_import(module_path: str, attribute: str) -> GraphFactory:
    """Return a wrapper that imports the factory on first call.

    This avoids loading all graph packages at startup — only the
    graph actually requested by the assistant is imported.
    """

    _cached: list[GraphFactory | None] = [None]

    async def _wrapper(config: Any, **kwargs: Any) -> Any:
        if _cached[0] is None:
            import importlib

            module = importlib.import_module(module_path)
            _cached[0] = getattr(module, attribute)
        return await _cached[0](config, **kwargs)  # type: ignore[misc]

    # Preserve a human-readable name for debugging.
    _wrapper.__qualname__ = f"lazy({module_path}.{attribute})"
    _wrapper.__name__ = attribute
    return _wrapper


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def register_graph(
    graph_id: str,
    factory: GraphFactory | None = None,
    *,
    module_path: str | None = None,
    attribute: str = "graph",
) -> None:
    """Register a graph factory under a given ``graph_id``.

    You can either pass a callable directly or provide a module path
    for lazy importing (recommended for built-in graphs to keep startup
    fast).

    Args:
        graph_id: Unique identifier (e.g. ``"agent"``,
            ``"research_agent"``).
        factory: An async callable with signature
            ``(config, *, checkpointer=None, store=None) -> compiled``.
            Mutually exclusive with *module_path*.
        module_path: Dotted module path for lazy import (e.g.
            ``"graphs.react_agent"``).  Used with *attribute*.
        attribute: Name of the factory function inside *module_path*
            (default ``"graph"``).

    Raises:
        ValueError: If neither *factory* nor *module_path* is provided,
            or if both are provided.

    Example::

        # Eager registration (factory already imported)
        register_graph("my_graph", my_graph_factory)

        # Lazy registration (imported on first use)
        register_graph("my_graph", module_path="graphs.my_graph", attribute="graph")
    """
    if factory is not None and module_path is not None:
        raise ValueError(
            f"register_graph({graph_id!r}): provide either 'factory' or "
            "'module_path', not both"
        )

    if factory is not None:
        _GRAPH_REGISTRY[graph_id] = factory
        logger.debug("Registered graph: %s (eager)", graph_id)
    elif module_path is not None:
        _GRAPH_REGISTRY[graph_id] = _lazy_import(module_path, attribute)
        logger.debug(
            "Registered graph: %s (lazy: %s.%s)", graph_id, module_path, attribute
        )
    else:
        raise ValueError(
            f"register_graph({graph_id!r}): provide either 'factory' or 'module_path'"
        )


def resolve_graph_factory(graph_id: str | None = None) -> GraphFactory:
    """Resolve a graph factory function from a ``graph_id`` string.

    Args:
        graph_id: The assistant's ``graph_id`` field.  ``None`` and
            unrecognised values fall back to ``"agent"``
            (:mod:`graphs.react_agent`).

    Returns:
        An async callable with signature
        ``(config, *, checkpointer=None, store=None) -> compiled_graph``.

    Example::

        factory = resolve_graph_factory("research_agent")
        agent = await factory(config, checkpointer=cp, store=st)
    """
    effective_id = graph_id or DEFAULT_GRAPH_ID

    factory = _GRAPH_REGISTRY.get(effective_id)
    if factory is not None:
        logger.debug("resolve_graph_factory: graph_id=%s → found", effective_id)
        return factory

    if effective_id != DEFAULT_GRAPH_ID:
        logger.warning(
            "resolve_graph_factory: unknown graph_id=%r — falling back to '%s'",
            effective_id,
            DEFAULT_GRAPH_ID,
        )
        factory = _GRAPH_REGISTRY.get(DEFAULT_GRAPH_ID)
        if factory is not None:
            return factory

    # Last resort: if registry is empty (shouldn't happen), import directly.
    logger.warning(
        "resolve_graph_factory: registry has no '%s' entry — importing directly",
        DEFAULT_GRAPH_ID,
    )
    from graphs.react_agent import graph as react_agent_graph

    return react_agent_graph


def get_available_graph_ids() -> list[str]:
    """Return the list of all registered graph IDs.

    Useful for the ``/info`` endpoint, assistant validation, and
    debugging.

    Returns:
        Sorted list of graph ID strings.
    """
    return sorted(_GRAPH_REGISTRY.keys())


# ---------------------------------------------------------------------------
# Built-in graph registration (lazy imports — no packages loaded at startup)
# ---------------------------------------------------------------------------

register_graph("agent", module_path="graphs.react_agent", attribute="graph")
register_graph("research_agent", module_path="graphs.research_agent", attribute="graph")
